In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/consume-complaints-dataset-fo-nlp/complaints_processed.csv


In [2]:
!git clone https://github.com/victoriamuinde07/NPL-Customer-Complaint-Project.git

Cloning into 'NPL-Customer-Complaint-Project'...
remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 9 (delta 1), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (9/9), 6.06 KiB | 3.03 MiB/s, done.
Resolving deltas: 100% (1/1), done.


### Bank Customer Complaints Classification

### Executive Summary.
This project leverages Natural Language Processing(NLP) to streamline customer complaint classification,aiming to reduce the time and complexity customers face when submitting complaints.By implementing an NLP model that automatically categorizes complaints based on content,the system eliminates the need for customers to navigate through numerous questions or options.The project's primary goals are to minimize the time spent on logging complaints and to enhance user experience with a simple,intuitive interface.

To achieve these goals,the project trained and evaluated various machine learning models,including Multinomial Naive Bayes,Support Vector Machine(SVM),Logistic Regression,Random Forest,and an ensembe model(ExtraTrees),before training and selecting a transformer-based model,BERT.The BERT model excelled in handling language tasks,achieving a Macro F1-score of 0.85,weighted F1-score of 0.89,and an accuracy of 89%,showing strong performance across all complaint categories.The focus on Macro F1-score ensured balanced classification

### Data Exploration

In [3]:
#importing necessary libraries.
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import re

from wordcloud import WordCloud
from collections import Counter

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import label_binarize
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.preprocessing import MinMaxScaler
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report,accuracy_score
from sklearn.svm import LinearSVC
from sklearn.metrics import confusion_matrix,roc_curve,roc_auc_score,auc
from sklearn.metrics import precision_recall_fscore_support
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import ExtraTreesClassifier

from imblearn.over_sampling import SMOTE
import joblib





In [4]:
df = pd.read_csv('/kaggle/input/consume-complaints-dataset-fo-nlp/complaints_processed.csv')
df.head(20)

Unnamed: 0              product  \
0            0          credit_card   
1            1          credit_card   
2            2       retail_banking   
3            3     credit_reporting   
4            4     credit_reporting   
5            5     credit_reporting   
6            6     credit_reporting   
7            7     credit_reporting   
8            8     credit_reporting   
9            9     credit_reporting   
10          10  mortgages_and_loans   
11          11          credit_card   
12          12          credit_card   
13          13     credit_reporting   
14          14      debt_collection   
15          15  mortgages_and_loans   
16          16  mortgages_and_loans   
17          17  mortgages_and_loans   
18          18  mortgages_and_loans   
19          19     credit_reporting   

                                            narrative  
0   purchase order day shipping amount receive pro...  
1   forwarded message date tue subject please inve...  
2   forwarded message cc sent friday pdt subject f...  
3   payment history missing credit report speciali...  
4   payment history missing credit report made mis...  
5   payment history missing credit report made mis...  
6   va date complaint experian credit bureau invol...  
7   account reported abbreviated name full name se...  
8   account reported abbreviated name full name se...  
9   usdoexxxx account reported abbreviated name fu...  
10  beginning mortgage held mb financial mb mortga...  
11  called request new york state covid relief pla...  
12  capital one secured credit account opened name...  
13  collection account acct opened balance account...  
14  collection agency system inc mn phone sent let...  
15  current loan provident funding applied refinan...  
16  current loan got deferment paid back credit hi...  
17  closed mortgage dovenmuhle mortgage paid full ...  
18  contacted bank america boa mortgage customer s...  
19  xxxxxxxx credit card year disputed charge vend...

In [5]:
# Displaying basic information about the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 162421 entries, 0 to 162420
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  162421 non-null  int64 
 1   product     162421 non-null  object
 2   narrative   162411 non-null  object
dtypes: int64(1), object(2)
memory usage: 3.7+ MB


The dataset has 162421 entries and three columns.
Unnamed:0 Appears to be an index column and can be ignored or removed.
Product-Contains the complaint category(target variable) with values like credit_card,retail_banking,credit_reporting etc.
#### There are 10 missing entries in the narrative column

In [6]:
# dropping the unnecessary 'Unnamed: 0' column.
df = df.drop(columns=['Unnamed: 0'])
# dropping rows with missing values in the narrative column
df = df.dropna(subset=['narrative'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 162411 entries, 0 to 162420
Data columns (total 2 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   product    162411 non-null  object
 1   narrative  162411 non-null  object
dtypes: object(2)
memory usage: 3.7+ MB


Unnecessary Unnamed: 0 column dropped and missing values in narrative column handled